dlt/destinations/path_utils.py

```diff
@@ -1,45 +1,198 @@
+    "H",  # 0-23
+    # Minutes
+    "mm",  # 00-59
+    "m",  # 0-59
+    # Days of week
+    "dddd",  # Monday, Tuesday, Wednesday
+    "ddd",  # Mon, Tue, Wed
+    "dd",  # Mo, Tu, We
+    "d",  # 0-6
+    # Quarters of the year
+    "Q",  # 1, 2, 3, 4
+}
+
+SUPPORTED_PLACEHOLDERS = DATETIME_PLACEHOLDERS.union(
+    {
+        "schema_name",
+        "table_name",
+        "load_id",
+        "file_id",
+        "ext",
+        "curr_date",
+        "timestamp",
+    }
+)
 
-# TODO: ensure layout only has supported placeholders
-SUPPORTED_PLACEHOLDERS = {"schema_name", "table_name", "load_id", "file_id", "ext", "curr_date"}
 
 SUPPORTED_TABLE_NAME_PREFIX_PLACEHOLDERS = ("schema_name",)
 
 
-def check_layout(layout: str) -> List[str]:
+def get_placeholders(layout: str) -> List[str]:
+    return re.findall(r"\{(.*?)\}", layout)
+
+
+def prepare_datetime_params(
+    current_datetime: Optional[pendulum.DateTime] = None,
+    load_package_timestamp: Optional[str] = None,
+) -> Dict[str, str]:
+    # For formatting options please see
+    # https://github.com/sdispater/pendulum/blob/master/docs/docs/string_formatting.md
+    # Format curr_date datetime according to given format
+    params: Dict[str, str] = {}
+    now: pendulum.DateTime = current_datetime or pendulum.now()
+    if load_package_timestamp:
+        package_datetime = pendulum.parse(load_package_timestamp)
+        params["timestamp"] = str(int(package_datetime.timestamp()))  # type: ignore[union-attr]
+
+    # Timestamp placeholder
+
+    # Take date from timestamp as curr_date
+    params["curr_date"] = str(now.date())
+
+    for format_string in DATETIME_PLACEHOLDERS:
+        params[format_string] = now.format(format_string).lower()
+
+    return params
+
+
+def prepare_params(
+    extra_placeholders: Optional[Dict[str, Any]] = None,
+    job_info: Optional[ParsedLoadJobFileName] = None,
+    schema_name: Optional[str] = None,
+    load_id: Optional[str] = None,
+) -> Dict[str, Any]:
+    params: Dict[str, Any] = {}
+    table_name = None
+    file_id = None
+    ext = None
+    if job_info:
+        table_name = job_info.table_name
+        file_id = job_info.file_id
+        ext = job_info.file_format
+        params.update(
+            {
+                "table_name": table_name,
+                "file_id": file_id,
+                "ext": ext,
+            }
+        )
+
+    if schema_name:
+        params["schema_name"] = schema_name
+
+    if load_id:
+        params["load_id"] = load_id
+
+    # Resolve extra placeholders
+    if extra_placeholders:
+        for key, value in extra_placeholders.items():
+            if callable(value):
+                try:
+                    params[key] = value(
+                        schema_name,
+                        table_name,
+                        load_id,
+                        file_id,
+                        ext,
+                    )
+                except TypeError:
+                    fmt.secho(
```

Comment:
wrap this in exception and raise from typeError

Reply:
Got it

Reply:
@rudolfix I added a new exception for this.